// @mui/x-internal-gestures/core/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "calculateAverageDistance", {
  enumerable: true,
  get: function () {
    return _calculateAverageDistance.calculateAverageDistance;
  }
});
Object.defineProperty(exports, "calculateCentroid", {
  enumerable: true,
  get: function () {
    return _calculateCentroid.calculateCentroid;
  }
});
Object.defineProperty(exports, "calculateRotationAngle", {
  enumerable: true,
  get: function () {
    return _calculateRotationAngle.calculateRotationAngle;
  }
});
Object.defineProperty(exports, "createEventName", {
  enumerable: true,
  get: function () {
    return _createEventName.createEventName;
  }
});
Object.defineProperty(exports, "getAngle", {
  enumerable: true,
  get: function () {
    return _getAngle.getAngle;
  }
});
Object.defineProperty(exports, "getDirection", {
  enumerable: true,
  get: function () {
    return _getDirection.getDirection;
  }
});
Object.defineProperty(exports, "getDistance", {
  enumerable: true,
  get: function () {
    return _getDistance.getDistance;
  }
});
Object.defineProperty(exports, "getPinchDirection", {
  enumerable: true,
  get: function () {
    return _getPinchDirection.getPinchDirection;
  }
});
Object.defineProperty(exports, "getVelocity", {
  enumerable: true,
  get: function () {
    return _getVelocity.getVelocity;
  }
});
Object.defineProperty(exports, "isDirectionAllowed", {
  enumerable: true,
  get: function () {
    return _isDirectionAllowed.isDirectionAllowed;
  }
});
var _calculateAverageDistance = require("./calculateAverageDistance");
var _calculateCentroid = require("./calculateCentroid");
var _calculateRotationAngle = require("./calculateRotationAngle");
var _createEventName = require("./createEventName");
var _getAngle = require("./getAngle");
var _getDirection = require("./getDirection");
var _getDistance = require("./getDistance");
var _getVelocity = require("./getVelocity");
var _isDirectionAllowed = require("./isDirectionAllowed");
var _getPinchDirection = require("./getPinchDirection");
